### ISSUE TITLE
Double event notification

### ISSUE THREAD

**Issue by pwinston** [napari/napari#2011]
## 🐛 Bug

This is subtle, but I think wrong. Although maybe it's debatable.

With octree code I connect to the root `camera.events` by calling `camera.events.connect(self._on_camera)`.  What I'm seeing is `_on_camera` is called at intervals like this, in milliseconds:

    1.8, 20, 1.7, 25, 1.6, 19, 1.4

So it's called twice each "frame" with 1-2ms between the calls.

I assume this is coming from here:

        with self._camera.events.center.blocker(self._on_angles_change):
            self._camera.angles = self.angles
        with self._camera.events.center.blocker(self._on_center_change):
            self._camera.center = self.center
        with self._camera.events.zoom.blocker(self._on_zoom_change):
            self._camera.zoom = self.zoom

I think when the camera moves my `_on_camera` is called twice, once because of `center` and once because of `zoom`. 

## Expected behavior

The camera only moved once, so ideally the "root" event handler should be called once.

This might be a general issue with "root" events, not with the camera. So here we have 3 leaf events, and two hit at once, so we call the root event twice. But what if there were 10 leaf events, and something caused all of them to fire at once. Then the root event would be called 10 times in a row. That seems clearly wrong?

There is some analogy here with action potentials maybe.

# Workaround

I am working around this with a filter. So my `_on_camera` will ignore being called "too quickly". This is 100% effective so I'm not waiting on this fix.

Note that I do want to be called for either situation (center or zoom). I just don't want to be called twice if one movement causes both to happen.

**Comment by sofroniewn** [comment]
> I think when the camera moves my _on_camera is called twice, once because of center and once because of zoom.

Yeah, that could be it, thanks for flagging. Agreed we should do something, I can think of a couple different options. 

> I am working around this with a filter. So my _on_camera will ignore being called "too quickly". This is 100% effective so I'm not waiting on this fix.

Good to know it's not blocking you right now.

>     with self._camera.events.center.blocker(self._on_angles_change):

That also looks like a typo bug - we should be calling the `self._camera.events.angles.blocker` instead

**Comment by pwinston** [comment]
Maybe you could fix it for the camera with a new event called "movement" or something. And that fires once and only once. And you just don't use the root event for the camera.

Fixing it "in general" for all events seems like a big can of worms. And people might be counting on the old way. And it's kind of debatable which is right. And it's Vispy's code really.

It does remind me of something that comes up a lot with real-time systems which are extensive use of dirty bits. So instead of A causes B right then, A just makes a note that B should happen. Then at some well-known time you do B if you are supposed to.

The advantage is B only happens once, no matter how many times A happened. But also regardless of when A happened, B happens at the same time every frame. Instead of B happening at all sorts of weird times.

An example in a game might be you update/move all the characters, and then you update any terrain square that had at least one character it. So no double updates, but also it clearly divides the frame into "updating characters" and "updating terrain" instead of a complex interleaving of the two.

**Comment by willingc** [comment]
I'm triaging old issues. I'm going to close this one out. I've added the notification label so it will surface when future work on notifications occur.
